### PR TITLE
Add fallback routines for empty secret cases

### PR DESCRIPTION
**Pull request by juliusknorr**
Make sure to keep authentication working when an instance has been setup without a secret after adding the secret manually afterwards.

- Password auth to webdav still works
- App password auth to webdav still works
- Browser session keeps being active
- PublicKeyTokens get rotated if the fallback is hit so their private key is reencrypted with the secret now

Provides a possible migration path for https://github.com/nextcloud/server/pull/31492

## ToDo
- [ ] Add migration step to update the config if possible and set a secret before the upgrade
  - [ ] secret
  - [ ] paswordsalt

### CODE REVIEW

File: lib/private/Authentication/Token/PublicKeyTokenProvider.php [hunk 1]
```diff
@@ -326,6 +332,10 @@ private function hashToken(string $token): string {
 		return hash('sha512', $token . $secret);
 	}
 
+	private function hashTokenWithEmptySecret(string $token): string {
```
Comment (nickvergessen) [inline lib/private/Authentication/Token/PublicKeyTokenProvider.php:335]: Please add `@deprecated` and explain the legacy reason

File: lib/private/Authentication/Token/PublicKeyTokenProvider.php [hunk 2]
```diff
@@ -301,9 +302,14 @@ private function decrypt(string $cipherText, string $token): string {
 		try {
 			return $this->crypto->decrypt($cipherText, $token . $secret);
 		} catch (\Exception $ex) {
-			// Delete the invalid token
-			$this->invalidateToken($token);
-			throw new InvalidTokenException("Could not decrypt token password: " . $ex->getMessage(), 0, $ex);
+			// Retry with empty secret as a fallback for instances where this might not have been set
```
Comment (nickvergessen) [inline lib/private/Authentication/Token/PublicKeyTokenProvider.php:305]:
```suggestion
			// Retry with empty secret as a fallback for instances where the secret might not have been set by accident
```

File: lib/private/Security/Crypto.php [hunk 3]
```diff
@@ -125,6 +125,16 @@ public function decrypt(string $authenticatedCiphertext, string $password = ''):
 		if ($password === '') {
 			$password = $this->config->getSystemValue('secret');
 		}
+		try {
+			return $this->decryptWithoutSecret($authenticatedCiphertext, $password);
+		} catch (Exception $e) {
+			// Retry with empty secret as a fallback for instances where this might not have been set
```
Comment (nickvergessen) [inline lib/private/Security/Crypto.php:131]:
```suggestion
			// Retry with empty secret as a fallback for instances where the secret might not have been set by accident
```

File: lib/private/Security/VerificationToken/VerificationToken.php [hunk 4]
```diff
@@ -84,7 +84,12 @@ public function check(string $token, ?IUser $user, string $subject, string $pass
 		try {
 			$decryptedToken = $this->crypto->decrypt($encryptedToken, $passwordPrefix.$this->config->getSystemValue('secret'));
 		} catch (\Exception $e) {
-			$this->throwInvalidTokenException(InvalidTokenException::TOKEN_DECRYPTION_ERROR);
+			// Retry with empty secret as a fallback for instances where this might not have been set
```
Comment (nickvergessen) [inline lib/private/Security/VerificationToken/VerificationToken.php:87]:
```suggestion
			// Retry with empty secret as a fallback for instances where the secret might not have been set by accident
```

File: lib/private/Security/Crypto.php [hunk 5]
```diff
@@ -125,6 +125,16 @@ public function decrypt(string $authenticatedCiphertext, string $password = ''):
...
+			return $this->decryptWithoutSecret($authenticatedCiphertext, '');
+		}
+
```
Comment (nickvergessen) [inline lib/private/Security/Crypto.php:134]:
```suggestion
```

**Comment by juliushaertl** [comment]
Test results when having an instance without a secret and adding one afterwards:
- Users get logged out of their sessions in the browser
- Reauthentication with existing password works -> password salt change is confirmed to not be an issue
- App passwords still fail

**Comment by juliushaertl** [comment]
Retested with follow up commits:

- [x] Password auth to webdav still works
- [x] App password auth to webdav still works
- [x] Browser session keeps being active
- [x] PublicKeyTokens get rotated if the fallback is hit so their private key is reencrypted with the secret now

**Review by CarlSchwan — APPROVED** [review]
Code looks good and I tested it manually and I didn't get logged when moving from a config without secret to a config.php with a secret

File: lib/private/Authentication/Token/PublicKeyTokenProvider.php [hunk 6]
```diff
@@ -326,6 +338,13 @@ private function hashToken(string $token): string {
 		return hash('sha512', $token . $secret);
 	}
 
+	/**
+	 * @depreacted Fallback for instances where the secret might not have been set by accident
```
Comment (nickvergessen) [inline lib/private/Authentication/Token/PublicKeyTokenProvider.php:342]:
```suggestion
	 * @deprecated Fallback for instances where the secret might not have been set by accident
```

**Comment by CarlSchwan** [comment]
Should we merge it without the migration path and just add the warning from https://github.com/nextcloud/server/pull/31492 instead?

It's better than the current state

**Review by ArtificialOwl — APPROVED** [review]

**Review by come-nc — APPROVED** [review]

**Comment by CarlSchwan** [comment]
Manual backport https://github.com/nextcloud/server/pull/35605

**Pull request merged**